### PR TITLE
timeutil: Throw parsing error in case of wrong Julian date format

### LINKED ISSUE #63250: sql: timezone() internal error: parsing as type timestamp: field Year is not wanted in 4088
Found on the current `master` by random syntax tests:

To reproduce:
```
SELECT timezone(e'S\r]':::STRING, e'\'j7\x17J3\x10\n\x1a':::STRING)
```

It causes an internal error, posted below. Instead we should probably have some sort of "invalid timestamp" error.
 
 ```
 E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +timezone(): parsing as type timestamp: field Year is not wanted in 4088
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +(1)
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +Wraps: (2) keys: [timezone()]
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +Wraps: (3) attached stack trace
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  -- stack trace:
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).MaybeWrapError
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:4026
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:4047
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:156
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:499
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:119
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:296
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:86
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:50
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.startExec
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:502
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:121

E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +Wraps: (4) timezone()
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +Wraps: (5) candidate pg code: 22007
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +Wraps: (6) attached stack trace
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  -- stack trace:
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/util/timeutil/pgdate.parseError
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/util/timeutil/pgdate/parsing.go:272
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/util/timeutil/pgdate.ParseTimestamp
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/util/timeutil/pgdate/parsing.go:212
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.ParseDTimestampTZ
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2427
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func154
E210407 20:04:25.584284 2213210 sql/sqltelemetry/report.go:57  [n1,client=127.0.0.1:46030,hostssl,user=root] 2 +  | 	/home/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:2628
```

### PR DESCRIPTION
**Pull request by kakshay21**
fixes: #63250
Release note (sql change): Fix Julian date parsing logic for wrong formats.

### CODE REVIEW

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 1]
```diff
@@ -765,7 +765,7 @@ func (fe *fieldExtract) Reset(field field, v int) error {
 // the field has already been set.
 func (fe *fieldExtract) Set(field field, v int) error {
 	if !fe.wanted.Has(field) {
-		return errors.AssertionFailedf("field %s is not wanted in %v", errors.Safe(field.Pretty()), errors.Safe(fe.wanted))
+		return inputErrorf("field %s is not wanted in %v", errors.Safe(field.Pretty()), errors.Safe(fe.wanted))
```
Comment (knz) [inline pkg/util/timeutil/pgdate/field_extract.go:768]: Can you avoid errors.Safe here

1) in the user-provided value, which is definitely unsafe

2) for the field name, instead have the field implement a SafeFormat method, or have a SafePretty method that returns redact.RedactableString

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63327)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/util/timeutil/pgdate/field_extract.go, line 768 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63327#-MXmhTPgiqTA8n3nV3Ky-r1-768:-MXqnw2T8C94o7ZrbS6t:bsm6lmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9b9d9ca1b1c9c156fc031ff5ff69166d72db4a/pkg/util/timeutil/pgdate/field_extract.go#L768)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you avoid errors.Safe here

1) in the user-provided value, which is definitely unsafe

2) for the field name, instead have the field implement a SafeFormat method, or have a SafePretty method that returns redact.RedactableString
</blockquote></details>

Ok so I have re-read the PR and I apologize for misleading you: i did not notice `errors.Safe` was already present in the code before your change.

(We need to fix this, but it is not your problem to fix)

So I think your first fix was OK. You can change it like that then I can approve the PR.

**Comment by kakshay21** [comment]
> So I think your first fix was OK. You can change it like that then I can approve the PR.
Sure no worries

**Comment by kakshay21** [comment]
Hi @knz not sure if `Bazel CI (Optional)` is falling because of my change

**Review by knz — APPROVED** [review]
Hi Kumar,

I looked at the PR again and we are going to keep your changes. The CI failure is not your responsibility (I will look into it).

Additionally, you will notice that I have added some commits to your branch on top of your change. The goal of these commits is to generally improve error generation in that area of the code. I hope you don't mind and I hope you get a chance to learn from my additions.

Cheers

> Reviewed 1 of 2 files at r1, 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63327#-:-MY5_9Z9BE6LPTW1ftym:b-glfteg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)

**Comment by knz** [comment]
(now waiting for CI success before merging)

**Comment by knz** [comment]
@rafiss PTAL, can you review my changes on top of the contributor's

File: pkg/util/timeutil/pgdate/field_extract_test.go [hunk 2]
```diff
@@ -107,6 +113,50 @@ func TestExtractSentinels(t *testing.T) {
...
+		}
+		err := fe.Extract(tc.str)
+		if err == nil {
+			t.Errorf("%+v: expected error, got no error", tc)
+			continue
+		}
+
+		msg := err.Error()
+		msg += "\n" + errors.FlattenDetails(err)
+		if msg != tc.expError {
```
Comment (rafiss) [inline pkg/util/timeutil/pgdate/field_extract_test.go:152]: are the error messages here stable enough be equality-checked in unit tests?

**Review by rafiss — APPROVED** [review]
thank you both! i left a question, but this looks good

File: pkg/util/timeutil/pgdate/field_extract_test.go (hunk 2 above)
Comment (knz) [reply pkg/util/timeutil/pgdate/field_extract_test.go:152]: in this case, yes - the full text of the message is generated within this package.
(I wouldn't be comfortable including bits of messages coming from other packages)

**Comment by knz** [comment]
bors r=rafiss

**Comment by kakshay21** [comment]
Thank you @knz and @rafiss

**Pull request merged**